Brieft on the Dataset : Mr.Menon is the owner of departmental stores he is been working with a analytics team on improving his buisness . The analytics team took the dataset of the past 2 months . The dataset consists of last items pucrhased aross 8000+ bills in the last 2 months

Buisness Problem being worked on : The analytics team has been keenly working on identifying patterns from the data .On what set of items combination can be placed adjacent to each other

Approach :  <br>
Importing essntial libararies <br>
Importing the excel file <br>
Creating a dummy dataframe to binary encode each item in the dataset <br>
The Items are then being sorted according to frequency of the items  <br>
The sorting is being done to identify the support of Items <br>
We can get an idea on what could be the possible threshold support value  <br>
We apply appriori algorithm with a threshold of a support value and extract the item sets <br>
Association rules can be applied on the identified itemset with a threshold lift value and possible combinations could be extratced <br>
The analytics team hands over the detailed working to Mr.Menon and his team and feasible arrangement inside the store can be done <br>




In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules
import pyfpgrowth as fptree

In [2]:
df_hotelbill=pd.read_excel("Grocery Stores.xlsx")

In [3]:
df_hotelbill.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9835 entries, 0 to 9834
Data columns (total 32 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   9835 non-null   object
 1   Unnamed: 1   7676 non-null   object
 2   Unnamed: 2   6033 non-null   object
 3   Unnamed: 3   4734 non-null   object
 4   Unnamed: 4   3729 non-null   object
 5   Unnamed: 5   2874 non-null   object
 6   Unnamed: 6   2229 non-null   object
 7   Unnamed: 7   1684 non-null   object
 8   Unnamed: 8   1246 non-null   object
 9   Unnamed: 9   896 non-null    object
 10  Unnamed: 10  650 non-null    object
 11  Unnamed: 11  468 non-null    object
 12  Unnamed: 12  351 non-null    object
 13  Unnamed: 13  273 non-null    object
 14  Unnamed: 14  196 non-null    object
 15  Unnamed: 15  141 non-null    object
 16  Unnamed: 16  95 non-null     object
 17  Unnamed: 17  66 non-null     object
 18  Unnamed: 18  52 non-null     object
 19  Unnamed: 19  38 non-null   

In [4]:
colnames= ['Item'+str(i) for i in range(1,33)]    
df_hotelbill.columns=colnames

In [5]:
item_set=set()
for i in range(df_hotelbill.shape[1]):
    for j in range(df_hotelbill.shape[0]):
        item_set.add(df_hotelbill.iloc[j,i])
    

In [6]:
item_set

{'Instant food products',
 'UHT-milk',
 'abrasive cleaner',
 'artif. sweetener',
 'baby cosmetics',
 'baby food',
 'bags',
 'baking powder',
 'bathroom cleaner',
 'beef',
 'berries',
 'beverages',
 'bottled beer',
 'bottled water',
 'brandy',
 'brown bread',
 'butter',
 'butter milk',
 'cake bar',
 'candles',
 'candy',
 'canned beer',
 'canned fish',
 'canned fruit',
 'canned vegetables',
 'cat food',
 'cereals',
 'chewing gum',
 'chicken',
 'chocolate',
 'chocolate marshmallow',
 'citrus fruit',
 'cleaner',
 'cling film/bags',
 'cocoa drinks',
 'coffee',
 'condensed milk',
 'cooking chocolate',
 'cookware',
 'cream',
 'cream cheese ',
 'curd',
 'curd cheese',
 'decalcifier',
 'dental care',
 'dessert',
 'detergent',
 'dish cleaner',
 'dishes',
 'dog food',
 'domestic eggs',
 'female sanitary products',
 'finished products',
 'fish',
 'flour',
 'flower (seeds)',
 'flower soil/fertilizer',
 'frankfurter',
 'frozen chicken',
 'frozen dessert',
 'frozen fish',
 'frozen fruits',
 'frozen m

In [7]:
df_hotelbill

Item1                Item2             Item3  \
0            citrus fruit  semi-finished bread         margarine   
1          tropical fruit               yogurt            coffee   
2              whole milk                  NaN               NaN   
3               pip fruit               yogurt     cream cheese    
4        other vegetables           whole milk    condensed milk   
...                   ...                  ...               ...   
9830              sausage              chicken              beef   
9831    cooking chocolate                  NaN               NaN   
9832              chicken         citrus fruit  other vegetables   
9833  semi-finished bread        bottled water              soda   
9834              chicken       tropical fruit  other vegetables   

                         Item4          Item5           Item6  \
0                  ready soups            NaN             NaN   
1                          NaN            NaN             NaN   
2                          NaN            NaN             NaN   
3                 meat spreads            NaN             NaN   
4     long life bakery product            NaN             NaN   
...                        ...            ...             ...   
9830            hamburger meat   citrus fruit          grapes   
9831                       NaN            NaN             NaN   
9832                    butter         yogurt  frozen dessert   
9833              bottled beer            NaN             NaN   
9834                   vinegar  shopping bags             NaN   

                Item7       Item8   Item9              Item10  ... Item23  \
0                 NaN         NaN     NaN                 NaN  ...    NaN   
1                 NaN         NaN     NaN                 NaN  ...    NaN   
2                 NaN         NaN     NaN                 NaN  ...    NaN   
3                 NaN         NaN     NaN                 NaN  ...    NaN   
4                 NaN         NaN     NaN                 NaN  ...    NaN   
...               ...         ...     ...                 ...  ...    ...   
9830  root vegetables  whole milk  butter  whipped/sour cream  ...    NaN   
9831              NaN         NaN     NaN                 NaN  ...    NaN   
9832    domestic eggs  rolls/buns     rum     cling film/bags  ...    NaN   
9833              NaN         NaN     NaN                 NaN  ...    NaN   
9834              NaN         NaN     NaN                 NaN  ...    NaN   

     Item24 Item25 Item26 Item27 Item28 Item29 Item30 Item31 Item32  
0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
1       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
2       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
3       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
4       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
...     ...    ...    ...    ...    ...    ...    ...    ...    ...  
9830    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
9831    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
9832    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
9833    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  
9834    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN  

[9835 rows x 32 columns]

In [8]:
dummy=np.zeros((df_hotelbill.shape[0],len(item_set)))

In [9]:
df_groecerydummy=pd.DataFrame(dummy)

In [10]:
df_groecerydummy.shape

(9835, 170)

In [11]:
df_groecerydummy.columns=item_set

In [12]:
df_groecerydummy

butter  NaN  cat food  packaged fruit/vegetables  newspapers  tea  \
0        0.0  0.0       0.0                        0.0         0.0  0.0   
1        0.0  0.0       0.0                        0.0         0.0  0.0   
2        0.0  0.0       0.0                        0.0         0.0  0.0   
3        0.0  0.0       0.0                        0.0         0.0  0.0   
4        0.0  0.0       0.0                        0.0         0.0  0.0   
...      ...  ...       ...                        ...         ...  ...   
9830     0.0  0.0       0.0                        0.0         0.0  0.0   
9831     0.0  0.0       0.0                        0.0         0.0  0.0   
9832     0.0  0.0       0.0                        0.0         0.0  0.0   
9833     0.0  0.0       0.0                        0.0         0.0  0.0   
9834     0.0  0.0       0.0                        0.0         0.0  0.0   

      pastry  processed cheese  nuts/prunes  kitchen towels  ...  pip fruit  \
0        0.0               0.0          0.0             0.0  ...        0.0   
1        0.0               0.0          0.0             0.0  ...        0.0   
2        0.0               0.0          0.0             0.0  ...        0.0   
3        0.0               0.0          0.0             0.0  ...        0.0   
4        0.0               0.0          0.0             0.0  ...        0.0   
...      ...               ...          ...             ...  ...        ...   
9830     0.0               0.0          0.0             0.0  ...        0.0   
9831     0.0               0.0          0.0             0.0  ...        0.0   
9832     0.0               0.0          0.0             0.0  ...        0.0   
9833     0.0               0.0          0.0             0.0  ...        0.0   
9834     0.0               0.0          0.0             0.0  ...        0.0   

      sausage  yogurt  frozen chicken  snack products  cooking chocolate  \
0         0.0     0.0             0.0             0.0                0.0   
1         0.0     0.0             0.0             0.0                0.0   
2         0.0     0.0             0.0             0.0                0.0   
3         0.0     0.0             0.0             0.0                0.0   
4         0.0     0.0             0.0             0.0                0.0   
...       ...     ...             ...             ...                ...   
9830      0.0     0.0             0.0             0.0                0.0   
9831      0.0     0.0             0.0             0.0                0.0   
9832      0.0     0.0             0.0             0.0                0.0   
9833      0.0     0.0             0.0             0.0                0.0   
9834      0.0     0.0             0.0             0.0                0.0   

      baby food  grapes  frankfurter  baby cosmetics  
0           0.0     0.0          0.0             0.0  
1           0.0     0.0          0.0             0.0  
2           0.0     0.0          0.0             0.0  
3           0.0     0.0          0.0             0.0  
4           0.0     0.0          0.0             0.0  
...         ...     ...          ...             ...  
9830        0.0     0.0          0.0             0.0  
9831        0.0     0.0          0.0             0.0  
9832        0.0     0.0          0.0             0.0  
9833        0.0     0.0          0.0             0.0  
9834        0.0     0.0          0.0             0.0  

[9835 rows x 170 columns]

In [13]:
df_groecerydummy.columns

Index([                   'butter',                         nan,
                        'cat food', 'packaged fruit/vegetables',
                      'newspapers',                       'tea',
                          'pastry',          'processed cheese',
                     'nuts/prunes',            'kitchen towels',
       ...
                       'pip fruit',                   'sausage',
                          'yogurt',            'frozen chicken',
                  'snack products',         'cooking chocolate',
                       'baby food',                    'grapes',
                     'frankfurter',            'baby cosmetics'],
      dtype='object', length=170)

In [14]:
df_groecerydummy.drop(np.nan,axis=1,inplace=True)

In [15]:
df_groecerydummy.margarine

0       0.0
1       0.0
2       0.0
3       0.0
4       0.0
       ... 
9830    0.0
9831    0.0
9832    0.0
9833    0.0
9834    0.0
Name: margarine, Length: 9835, dtype: float64

In [16]:
for i in range(df_hotelbill.shape[0]):
    for j in range(df_hotelbill.shape[1]):
        if(pd.isna(df_hotelbill.iloc[i,j])==False):
            df_groecerydummy.loc[i,df_hotelbill.iloc[i,j]]=1

In [17]:
df_groecerydummy

butter  cat food  packaged fruit/vegetables  newspapers  tea  pastry  \
0        0.0       0.0                        0.0         0.0  0.0     0.0   
1        0.0       0.0                        0.0         0.0  0.0     0.0   
2        0.0       0.0                        0.0         0.0  0.0     0.0   
3        0.0       0.0                        0.0         0.0  0.0     0.0   
4        0.0       0.0                        0.0         0.0  0.0     0.0   
...      ...       ...                        ...         ...  ...     ...   
9830     1.0       0.0                        0.0         0.0  0.0     0.0   
9831     0.0       0.0                        0.0         0.0  0.0     0.0   
9832     1.0       0.0                        0.0         0.0  0.0     0.0   
9833     0.0       0.0                        0.0         0.0  0.0     0.0   
9834     0.0       0.0                        0.0         0.0  0.0     0.0   

      processed cheese  nuts/prunes  kitchen towels  cookware  ...  pip fruit  \
0                  0.0          0.0             0.0       0.0  ...        0.0   
1                  0.0          0.0             0.0       0.0  ...        0.0   
2                  0.0          0.0             0.0       0.0  ...        0.0   
3                  0.0          0.0             0.0       0.0  ...        1.0   
4                  0.0          0.0             0.0       0.0  ...        0.0   
...                ...          ...             ...       ...  ...        ...   
9830               0.0          0.0             0.0       0.0  ...        0.0   
9831               0.0          0.0             0.0       0.0  ...        0.0   
9832               0.0          0.0             0.0       0.0  ...        0.0   
9833               0.0          0.0             0.0       0.0  ...        0.0   
9834               0.0          0.0             0.0       0.0  ...        0.0   

      sausage  yogurt  frozen chicken  snack products  cooking chocolate  \
0         0.0     0.0             0.0             0.0                0.0   
1         0.0     1.0             0.0             0.0                0.0   
2         0.0     0.0             0.0             0.0                0.0   
3         0.0     1.0             0.0             0.0                0.0   
4         0.0     0.0             0.0             0.0                0.0   
...       ...     ...             ...             ...                ...   
9830      1.0     0.0             0.0             0.0                0.0   
9831      0.0     0.0             0.0             0.0                1.0   
9832      0.0     1.0             0.0             0.0                0.0   
9833      0.0     0.0             0.0             0.0                0.0   
9834      0.0     0.0             0.0             0.0                0.0   

      baby food  grapes  frankfurter  baby cosmetics  
0           0.0     0.0          0.0             0.0  
1           0.0     0.0          0.0             0.0  
2           0.0     0.0          0.0             0.0  
3           0.0     0.0          0.0             0.0  
4           0.0     0.0          0.0             0.0  
...         ...     ...          ...             ...  
9830        0.0     1.0          0.0             0.0  
9831        0.0     0.0          0.0             0.0  
9832        0.0     0.0          0.0             0.0  
9833        0.0     0.0          0.0             0.0  
9834        0.0     0.0          0.0             0.0  

[9835 rows x 169 columns]

In [18]:
total_items_sold= sum(df_groecerydummy.sum())

In [19]:
total_item_count

NameError: name 'total_item_count' is not defined

In [ ]:
item_frequency=pd.DataFrame(df_groecerydummy.sum().sort_values(ascending=False)).reset_index()

In [ ]:
item_frequency.columns

In [ ]:
item_frequency.rename(columns={item_frequency.columns[0]:'Item_Name',item_frequency.columns[1]:'Frequency'},inplace=True)

In [ ]:
item_frequency

In [ ]:
##Top 30 Sold Items

In [ ]:
top_30_items=list(item_frequency.Item_Name.head(30))

In [ ]:
top_30_items

In [ ]:
item_len= np.arange(len(top_30_items))
top_30_frequency = list(item_frequency['Frequency'].head(n=30))
plt.bar(item_len, top_30_frequency)
plt.xticks(y_pos, top_30_items, rotation='vertical')
plt.ylabel('Number of Goods Sold')
plt.title('Top 30 Items Frequency Distribution')

In [ ]:
plt.hist(item_frequency['Frequency'])

As per the above Histogram Projection there are many items which has frequency less than 250. So the minimum supoort criteria should be as minimum as possible 

In [ ]:
item_frequency.iloc[0,:].Frequency

In [ ]:
item_with_max_support=(item_frequency.iloc[0,:].Frequency)/df_groecerydummy.shape[0]
item_with_max_support

In [ ]:
item_with_min_support=(item_frequency.iloc[168,:].Frequency)/df_groecerydummy.shape[0]
item_with_min_support

Key Facts about the Dataset <br>
Maximum Support of an Item Sold : 0.255 <br>
Minimum Support of an Item Sold : 0.0001 <br>
Number of Unique Items : 169 <br>
Number of Bills : 9835

### Apriori Algorithm

In [ ]:
most_combined_items = apriori(df_groecerydummy,min_support=0.02,use_colnames=True)
rules = association_rules(most_combined_items, metric="lift", min_threshold=1.5)

In [ ]:
most_combined_items

In [ ]:
rules[['antecedents','consequents']]

### FP Growth Tree

In [ ]:
most_combined_items_fptree = fptree.find_frequent_patterns(df_groecerydummy,0.02)
rules_fptree = fptree.generate_association_rules(most_combined_items_fptree,1.5)

In [ ]:
rules_fptree

In [ ]:
rules.to_excel("FinalItemSet.xlsx")

Conclusion : From the above case study on Mr.Menon's Departmental Store. We did a market basket analysis on bill wise data . When we applied apriori algorithm to the grocery dataset we would get subsets of items which follow the given condition (Support >0.02) . Association rule is applied on the set of items extracted from apriori. The possible combination which has lift greater than 1.5. Is being considered as the good combonation. We can also change thresholds and extract outputs

Most Suggested Items sets are <br>
1)  Dairy Products and Vegetables <br>
2)  Meat and Dairy Products <br>
3)  Fruits and Dairy Products